lib/system/web/src/keyboard_engine.rs

```diff
@@ -0,0 +1,122 @@
+use std::collections::{HashMap, BTreeSet};
+
+use web_sys::{KeyboardEvent};
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+
+use crate::*;
+
+type Bindings = HashMap<BTreeSet<u32>, Vec<Box<dyn FnMut()>>>;
+
+/// Keyboard Engine that handles user keyboard combinations
+///
+/// Keyboard Engine stores bindings as HashMap:
+/// {
+///     [keyA, keyB]:          [Callback1, Callback2],
+///     [keyA, keyB, keyCtrl]: [Callback1],
+/// }
+/// when the combination is pressed related callbacks are invoked
+pub struct KeyboardEngine {
+    pub bindings: Rc<RefCell<Bindings>>,
+    pub state:    Rc<RefCell<BTreeSet<u32>>>,
+    pub target:   web_sys::EventTarget
```

Comment:
What if you don't keep `Rc<RefCell<...>>` on fields, but instead you keep ownership of all th fields and also you create another structure, like `KeyboardEngineShared (Rc<RefCell< KeyboardEngine>>)` instead? This way you can keep significant portion of the API as mutable and use interior mutability abstracted away.

Reply:
Doesn't it increase complexity? I need `RefCell` as I described above for being able to handle keyboard events. Moving this logic to another struct will lead that all the code we have right now appears in the another struct what creates a mess - why I need the `KeyboardEngineInternal` then. 
KeyboardEngine should be bounded with logic of subscribing on keyboard events, cause it's the core functionality and without it it loses the meaning.  
I tried different approaches and haven't found better way because of `WasmClosure` 

I've checked Servo source code and found that using `Cell` is really common there even for simple structs
[KeyboardEvent.rs](https://github.com/servo/servo/blob/8cb614531183335ea320f60cb2236c7f4ba64480/components/script/dom/keyboardevent.rs#L25)  
[Event.rs](https://github.com/servo/servo/blob/8cb614531183335ea320f60cb2236c7f4ba64480/components/script/dom/event.rs#L50)

What I did at this point was change `RefCell` to `Cell` for state